[Accessibility] Using arrow keys to navigate in toolbars items

brichet opened the pull request:
This PR allows to use the arrow keys to navigate in toolbars items.

https://github.com/jupyterlab/jupyterlab/assets/32258950/39f91d8f-d941-44a4-986a-e59f38c1fa37

The toolbars use a [toolkit](https://github.com/jupyterlab-contrib/jupyter-ui-toolkit) based on  web components. This is a (first ?) step of https://github.com/jupyterlab/team-compass/issues/143.

## Reference

####  About accessibility:

- https://github.com/jupyter/notebook/issues/6595
- https://github.com/jupyter/notebook/issues/6800
- https://github.com/jupyter/notebook/issues/6935
- https://github.com/jupyter/notebook/issues/6957
- https://github.com/jupyterlab/jupyterlab/issues/9399
- Fixes https://github.com/jupyterlab/jupyterlab/issues/12187


#### About web components:

- https://github.com/jupyterlab/team-compass/issues/143

## Code changes

Make use of the web component from [jupyter-ui-toolkit](https://github.com/jupyterlab-contrib/jupyter-ui-toolkit) to build the toolbars and the toolbar items.

## User-facing changes

EDIT:
~~The layout of the toolbars is not exactly the same as before. Let's wait for ui-tests to see the difference...~~
- toolbars are 3 pixels wider, to increase the border width when the items are focused
- toggled buttons has not the same appearance 

| | not toggled | toggled |
| ---  | --- | --- |
| before | <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/ae244d6f-cf58-4675-8992-86d231ebe41b width=30px> |<img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/0e08604b-158a-4334-83eb-b74a96e778a8 width=30px> |
| after | <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/8639c35b-82c9-4aa1-a646-6aebe561e391 width=30px> | <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/75fb3b24-0342-4a64-b81e-d13a36849899 width=30px>   |

## Backwards-incompatible changes

The toolbars built from the `ui-components` package should use the web components items from [jupyter-ui-toolkit](https://github.com/jupyterlab-contrib/jupyter-ui-toolkit).

cc. @fcollonval who contributed to this PR

brichet commented:
please update snapshots

brichet commented:
Bumping the CI

brichet reviewed (commented):
During the tests, sometime the text in toolbar is not visible. 
I had to add rules or to modify the tests for them to pass (see below).

For now I only saw this during the playwright tests.
It may be related to the `color: --neutral-foreground-rest;` coming from *fast-foundation* components.
Maybe we should try to understand why/when it occurs and if it has to be fixed.

brichet commented:
please update galata snapshots

brichet commented on galata/test/jupyterlab/debugger.test.ts:
```diff
@@ -64,6 +64,8 @@ test.describe('Debugger Tests', () => {
       'start-debug-session-sources.png'
     );
+
+    await page.click('jp-button[title^=Continue]');
```
I don't know why, but for this specific test we need to continue the execution, otherwise the `afterEach` function is failing.

brichet commented:
I opened https://github.com/jupyterlab/jupyterlab/issues/15116 for the browser_check fails

brichet commented:
please update documentation snapshots

brichet commented on examples/cell/index.css:
```diff
@@ -22,6 +22,10 @@ body {
 
+.jp-Toolbar-kernelName {
+  color: var(--jp-ui-font-color1);
+}
```
Fixed in https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/releases/tag/v0.13.1

brichet commented on galata/test/documentation/debugger.test.ts:
```diff
@@ -319,12 +319,9 @@ test.describe('Debugger', () => {
-    await expect(
-      page.locator(
-        '[aria-label="side panel content"] >> text=Source/tmp/ipykernel_'
-      )
-    ).toBeVisible();
+    await expect(page.locator('.jp-DebuggerSources-header-path')).toContainText(
+      '/tmp/ipykernel_'
+    );
```
Fixed in https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/releases/tag/v0.13.1

brichet commented:
There are still a few glitches that can be seen in the snapshots:

- galata/test/documentation/debugger.test.ts-snapshots/debugger-run-documentation-linux.png

  The focus is not on the expected button
  <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/a6c15665-fe38-4d80-82fd-ea9f4601f837 width=400px>
  
  Running playwright in debug mode, we can see the focus on an other button:
  
  <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/021804f6-8a63-42cf-aa37-755564d66598 width=400px>

- galata/test/jupyterlab/notebook-toolbar.test.ts-snapshots/run-cell-jupyterlab-linux.png

  The cell following the executed one is not focused, but it does not occurs outside of the tests

  <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/32845e29-2c6c-4bb1-bb0e-b6ba43750fcb>

- galata/test/documentation/export_notebook.test.ts-snapshots/exporting-slide-type-documentation-linux.png

  The notebook toolbar should be fully displayed, without the opener button.

  ![image](https://github.com/jupyterlab/jupyterlab/assets/32258950/e4ea748a-5222-4320-b436-63601f918592)

 There is a hidden button next to the spacer. Previously this hidden button had no width, so it was really hidden. With this PR, the button have now a width and makes the reactive toolbar computation fail.
  I also found an other hidden button one next to the `topBar` spacer.
  I have no clue why these buttons are there (they are not described in their toolbar's respective schema).

  <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/5ca3d21b-d0d4-493c-af47-d8d4c33f790d width=300px>

  <img src=https://github.com/jupyterlab/jupyterlab/assets/32258950/7d0323fa-ad78-43d3-ad5e-96ffbf5ce92b width=300px>

brichet commented:
> * The cell following the executed one is not focused, but it does not occurs outside of the tests

The focus is taken by the 'run' button, which is why it doesn't remain in the cell editor.

fcollonval commented:
The latest commit stop the propagation on click event if the action is on mousedown to avoid the toolbar to change the focus.

![toolbar_button_focus](https://github.com/jupyterlab/jupyterlab/assets/8435071/a01ac8fb-b0c8-497c-aee0-b16308767c66)

There is an side effect of that workaround: if after clicking on an item, the user focus into the toolbar through keyboard navigation. The focus won't be on the latest clicked item as the toolbar was not notified it needed to change its focused element.

krassowski commented:
The Binder does not seem to build, merging main branch might help (there were improvements merged recently)

brichet commented:
> The Binder does not seem to build, merging main branch might help (there were improvements merged recently)

@krassowski I just rebased.

The rebase broke the test on `notebook-mobile` locally, I'll fix it if it is still the case on CI.

fcollonval commented:
Hey @krassowski are you ok if we merge this one?

krassowski commented:
I will take a look tonight and merge if all good.

krassowski commented:
One more - cell type switcher is not usable when collapsed into the "more commands" section (I do love the styling in the dark theme though, it looks amazing);

![cell-type-switcher](https://github.com/jupyterlab/jupyterlab/assets/5832902/d5c76dd8-e7fc-47bc-a6aa-6dcef5b6da29)

brichet commented:
Thanks @krassowski for the review.


- > * there are >200 changed files; it would be make the review easier if this PR did not change the toolbar height but I will not block on it if you believe this is required

  IIRC this is to increase the size of the focus border, and make it more visible, without changing the size of the widget.

- > * when the interface switcher is present I cannot use keys to navigate toolbar items in notebook toolbar (also see below for another interface switcher issue); this seems not ideal if extensions/installation of Jupyter Notebook can break navigation

  I don't have the same switcher than you have in the screenshot. On my side this is only a button, not a dropdown, and it works well with it.

- > * this is a somewhat breaking change for extension users, this is what users with kernelspy and nbdime installed will see; we should (a) avoid such a change if targeting a minor release (b) if it absolutely has to happen, add an extension migration entry on it

  I can't remember why we replaced
  ```
  {props.label && (
    <span className="jp-ToolbarButtonComponent-label">{props.label}</span>
  )}
  ```
  with 
  ```
  {props.label ?? ''}
  ```
  This should probably be reverted.


- > * I get some odd, blank button between kernel text and busy indicator; sometimes it is after sometimes it is before the busy button

  I can't reproduce it for the Notebook toolbar.
  For the menu toolbar, the buttons also exist before this PR, but have no width. I don't know where they come from. We can probably set the width to 0 px for empty button, but I think it should be fixed upstream.

- > The "more commands" button sometimes is outside of the toolbar
  
  I can't reproduce it, do you have an idea when it occurs ?

krassowski commented:
> I don't have the same switcher than you have in the screenshot. 

I think you may need to have notebook 7 and nbclassic installed to see this in the switcher form.

> I can't reproduce it, do you have an idea when it occurs ?

I think it was appearing after reloading JupyterLab with limited space but I do not remember very well. Maybe there is something related to what happens when buttons from extensions get added with a delay (after resize events settled)? Or maybe something to do with the labels or disabled buttons? I saw that when running with extensions enabled in the dev mode (I had a number of extensions installed, including latest `nbdime` pre-release to testing the downstream consequences).

krassowski commented on packages/ui-components/src/components/toolbar.tsx:
```diff
@@ -672,29 +714,23 @@ export function ToolbarButtonComponent(
-          className="jp-ToolbarButtonComponent-icon"
-          tag="span"
-          stylesheet="toolbarButton"
+          tag={null}
```
Note: I tested with a third-party theme and it looked ok. I think that the consequences will be minimal, resolving.

brichet commented:
> I think you may need to have notebook 7 and nbclassic installed to see this in the switcher form.

Thanks, it was it. The switcher is a `MenuBar` widget (Lumino), and the `MenuBar` catches the left and right arrow keys event to switch from one menu item to another (in this case there is no other menu item). This behavior would probably break any tentative of using arrow keys in the toolbar. I think it should be replaced with a select in Notebook project, to make it compatible with accessibility needs.

krassowski commented:
>  I think it should be replaced with a select in Notebook project, to make it compatible with accessibility needs.

Thanks for investigating! Can you open an issue in the Notebook repo to track this as an item for Notebook 7.1?

brichet commented on galata/test/documentation/customization.test.ts-snapshots/default-notebook-toolbar-documentation-linux.png:
IIRC we didn't find a proper way to have a dynamic width on the select. Therefore, we choose to fix it width to the largest word, to avoid truncating the selection. This creates a space between selection and caret (when the selected option is small), which look awkward  without border. I'll have a look if there is a way to fix it and keep the previous styling.

brichet replied:
I finally restored the previous widget for the cell type dropdown.
The component we used was not working correctly, some options were hidden if the opener toolbar was visible (problem of z-index).
In addition, to correctly display the options, we had to allow the y-overflow on toolbar. Consequently, when the toolbar size was changed, the buttons where visible below the toolbar for some milliseconds before disappearing.

brichet commented on packages/running/style/base.css:
```diff
@@ -113,8 +113,7 @@
 
-.jp-RunningSessions-shutdownAll.jp-mod-styled
-  > .jp-ToolbarButtonComponent-label {
+.jp-RunningSessions-shutdownAll.jp-mod-styled.jp-ToolbarButtonComponent {
```
Removed in https://github.com/jupyterlab/jupyterlab/pull/15021/commits/6e48c44a8de9cbd2b717cc2157d604482f65f42d

brichet commented on packages/ui-components/src/components/toolbar.tsx:
```diff
@@ -450,83 +459,106 @@ export class ReactiveToolbar extends Toolbar<Widget> {
-        }
-        if (width > toolbarWidth) {
-          widgetsToRemove.push(widget);
+    if (!(this.parent && this.parent.isAttached)) {
```
Thanks for catching this @krassowski.

I notice that the widget position is not preserved when the size is changed quickly.
Is this what you mean with this comment ?

I think I know what's wrong: the `widgetsToRemove` list is now asynchronously retrieves, so the throttler has almost no effect on the `_onResize` function. This creates a confusion in the widget order. 

I'll work on it, it looks like a simple semaphore like *computationInProgress* fixes it, but adds delay and sometime do not display all the element in the toolbar.

brichet replied:
It should be fixed with https://github.com/jupyterlab/jupyterlab/pull/15021/commits/cfe7c00241e8b683a3d92e0e38e6f4f846273ad8

brichet commented:
> * I get some odd, blank button between kernel text and busy indicator; sometimes it is after sometimes it is before the busy button

These were the command buttons with an unregistered command. The settings are loaded even if the extension is not.
Thanks to @fcollonval, I fixed it in https://github.com/jupyterlab/jupyterlab/pull/15021/commits/0fc22a8b8c10b7f410e90e59e6303e1c2e8b79ee

brichet commented on galata/test/jupyterlab/notebook-run-mermaid.test.ts-snapshots/run-cells-mermaid-jupyterlab-linux.png:
I opened https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/pull/77 to remove the background.

brichet replied:
Should also be fixed with https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/pull/77

brichet commented on galata/test/jupyterlab/notebook-run-vega.test.ts-snapshots/run-cells-dark-vega-jupyterlab-linux.png:
I opened https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/pull/77 to remove the background.

brichet commented on packages/ui-components/src/icon/labicon.tsx:
```diff
@@ -332,28 +332,39 @@ export class LabIcon implements LabIcon.ILabIcon, VirtualElement.IRenderer {
+    }
 
-      returnSvgElement = false;
+    const svgElement = this.svgElement.cloneNode(true) as HTMLElement;
```
This statement was already present below (line 353), which should have no impact on performance.

brichet commented:
@krassowski I think I addressed all your comments, if you can take another look.

The contrast on disabled buttons will have to wait for https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/pull/77

krassowski commented on packages/filebrowser/test/browser.spec.ts:
```diff
@@ -75,7 +75,7 @@ describe('filebrowser/browser', () => {
       it('toolbar should have an aria label of file browser and a role of navigation', () => {
```
```suggestion
      it('toolbar should have an aria label of file browser and a role of toolbar', () => {
```

krassowski commented on packages/debugger/style/kernelSources.css:
```diff
@@ -34,3 +34,19 @@
+  background-color: var(--jp-layout-color2);
+}
+
+.jp-DebuggerKernelSource-source svg {
```
Please replace the bare use of `svg` with a selector or `>`

brichet replied:
Thanks @krassowski, I thought the selector checked the first element before the others.

Fixed in https://github.com/jupyterlab/jupyterlab/pull/15021/commits/31d2e47b92a0b72a6a440c1cd8c9c232ac43d8e7

krassowski commented on packages/completer/style/base.css:
```diff
@@ -329,7 +329,7 @@
 
-[data-command='inline-completer:next'] > .jp-ToolbarButtonComponent-icon,
-[data-command='inline-completer:previous'] > .jp-ToolbarButtonComponent-icon {
+.jp-InlineCompleter [data-command='inline-completer:next'] svg,
+.jp-InlineCompleter [data-command='inline-completer:previous'] svg {
```
Can you please avoid bare `svg` selectors? It has a sever performance penalty when notebooks include many svg visualisations. The selector needs to rewritten to use child combinator (`>`) or a class name instead.

brichet replied:
Fixed in https://github.com/jupyterlab/jupyterlab/pull/15021/commits/31d2e47b92a0b72a6a440c1cd8c9c232ac43d8e7

krassowski commented on packages/testing/src/jest-config.ts:
```diff
@@ -7,7 +7,11 @@ import path from 'path';
 
 const esModules = [
   '@codemirror',
+  '@microsoft',
+  '@jupyter/react-components',
+  '@jupyter/web-components',
   '@jupyter/ydoc',
+  'exenv-es6',
```
Highlighting this one again: should there be a mention of the new ones in the extension migration guide for 4.1?

brichet replied:
Updated in https://github.com/jupyterlab/jupyterlab/pull/15021/commits/2a19c6d61d1337504b819e13f9f4dc123f421f86

krassowski reviewed (commented):
Yet to check it in action, but I went over the code and had a few small suggestions.

brichet commented on galata/test/documentation/general.test.ts-snapshots/jupyterlab-documentation-linux.png:
The snapshot is the same here https://github.com/jupyterlab/jupyterlab/pull/14115/files#diff-c2c16a939dea4e658e5384f51ccb1501f09b6393d1e5e87ce67112e396317776. and I can see the welcome test fails on other PRs, probably for the same reason.

It should probably be fixed in a dedicated PR.

brichet commented on galata/test/documentation/customization.test.ts-snapshots/customized-notebook-toolbar-documentation-linux.png:
Fixed in https://github.com/jupyterlab/jupyterlab/commit/31d2e47b92a0b72a6a440c1cd8c9c232ac43d8e7

brichet commented:
please update snapshots

brichet commented:
please update snapshots

brichet commented:
Bumping the CI

fcollonval reviewed (approved):
thanks @brichet

krassowski commented on docs/source/extension/extension_migration.rst:
```diff
@@ -27,6 +27,32 @@ If your extension previously included a custom enable/disable setting, you may b
+The Toolbar and ToolbarButtonComponent (from the package *ui-components*) now relies on the external library
+`jupyter-ui-toolkit <https://github.com/jupyterlab-contrib/jupyter-ui-toolkit>`.
+
+This library uses of the web component technology (https://developer.mozilla.org/en-US/docs/Web/API/Web_components),
```
```suggestion
This library uses the web component technology (https://developer.mozilla.org/en-US/docs/Web/API/Web_components),
```

krassowski commented on galata/test/jupyterlab/html-viewer.test.ts:
```diff
@@ -46,7 +46,10 @@ test.describe('HTML Viewer', () => {
 
   test('should allow links for trusted file', async ({ page }) => {
-    await page.getByRole('button', { name: 'Trust HTML' }).click();
+    await page.pause();
```
Is there a reason why it was kept? As far as I understand it is a debug statement which only has effect in headed mode (https://playwright.dev/docs/api/class-page)

> Pauses script execution. Playwright will stop executing the script and wait for the user to either press 'Resume' button in the page overlay or to call playwright.resume() in the DevTools console.

krassowski commented on packages/notebook/style/toolbar.css:
```diff
@@ -37,25 +37,12 @@
-  z-index: 1;
-  right: 0;
-  top: 0;
+.jp-Notebook-toolbarCellTypeDropdown span {
+  top: 5px !important;
 }
```
I were going to suggest adding `>` but it appears that this rule is not even needed:

```suggestion
```

![Screenshot from 2023-11-17 18-53-37](https://github.com/jupyterlab/jupyterlab/assets/5832902/74dd7cd9-8cf4-4b29-93d3-58fdb1712f38)

brichet replied:
Thanks @krassowski, I removed this rule.

krassowski reviewed (approved):
Thank you @brichet!

> We can explore alternatives to the visual changes in this PR in follow-up issues/PR if anyone reading this has strong opinions against the increase in toolbar height/changes to the debugger buttons.
